Automattic/jetpack
# AI Featured Image: support backend prompts

The frontend code change to implement backend prompts on the image generation. Test in pair with D150640-code.

## Proposed changes:


* Change the image generation hook to export the generic image generation function
* Remove specific mentions to models on the AI Featured Image tool, so the model selection happens on the backend
* With this change, the backend will select the model for the generation; the D150640-code change will use Dall-e-3 for Jetpack sites and Stable Diffusion for Simple sites

Notes:

* the model detection for logging/tracking purposes is not perfect in this PR (it's using the cost of the request as a proxy, assuming each model has different costs). A better way would be to expose the "active model" or "default model" on the "ai-assistant-feature" endpoint, but this is a bigger change considering the timelime we have before the next release (need to add some code on the backend, as well as change the Jetpack [here](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/_inc/lib/class-jetpack-ai-helper.php#L352))
* I kept the original prompts (and methods) for the Dall-e and Stable Diffusion generation on the hook so we can keep referring to it; there is always time for some clean up in the future

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






Testing on a Jetpack site:

* Apply D150640-code to your sandbox and sandbox the `public-api`
* Run this branch on your Jetpack testing site
* Open the browser console and enable debug with `localStorage.setItem( 'debug', '*' )`
* Go to the block editor and add some content
* Trigger the featured image generation
* On the console, confirm the event tracking shows the model as `dall-e-3`:

<img width="300" alt="Screenshot 2024-06-02 at 11 22 53" src="https://github.com/Automattic/jetpack/assets/6760046/acc0f147-aee1-4144-b12c-9ca280067e35">

* Confirm the payload of the request does not mention the model of the generation, and contains the `messages` array:

```
{
    "feature": "featured-post-image",
    "size": "1792x1024",
    "responseFormat": "b64_json",
    "style": "photographic",
    "messages": [
        {
            "role": "jetpack-ai",
            "context": {
                "type": "featured-image-generation",
                "request": null,
                "content": "the content"
            }
        }
    ]
}
```

* Confirm the image was generated by Dall-e-3
   * the image will "looks like Dall-e"
   * if you inspect the image, you will see the dimensions are 1792×1024

<img width="600" alt="Screenshot 2024-06-02 at 11 22 53" src="https://github.com/Automattic/jetpack/assets/6760046/72240489-3f41-4f77-bf9a-a942c330dc24">

Testing on a Simple site:

* Apply D150640-code to your sandbox
* After it, apply this branch on your sandbox using the `jetpack-downloader` command provided below
   * the tool will ask you to checkout trunk before applying it, type `n` to stay on the current branch
   * the tool will ask for another confirmation, to stay on the current branch, type `y` to stay on it
* Sandbox the `public-api` and also your testing simple site
* Open the browser console and enable debug with `localStorage.setItem( 'debug', '*' )`
* Go to the block editor and add some content
* Trigger the featured image generation
* On the console, confirm the event tracking shows the model as `stable-diffusion`

<img width="300" alt="Screenshot 2024-06-02 at 11 18 18" src="https://github.com/Automattic/jetpack/assets/6760046/39314366-4b12-4777-975f-24c6a9023aa0">

* Confirm the payload of the request does not mention the model of the generation, and contains the `messages` array:

```
{
    "feature": "featured-post-image",
    "size": "1792x1024",
    "responseFormat": "b64_json",
    "style": "photographic",
    "messages": [
        {
            "role": "jetpack-ai",
            "context": {
                "type": "featured-image-generation",
                "request": null,
                "content": "the content"
            }
        }
    ]
}
```

* Confirm the image was generated by Stable Diffusion
   * the image will "looks like Stable Diffusion"
   * if you inspect the image, you will see the dimensions are 1024x768

<img width="600" alt="Screenshot 2024-06-02 at 11 18 18" src="https://github.com/Automattic/jetpack/assets/6760046/31921297-b7b4-4a1f-92bf-4d0d17050789">
